libs/elasticsearch/langchain_elasticsearch/cache.py

```diff
@@ -186,3 +236,213 @@ def clear(self, **kwargs: Any) -> None:
+        Collapse the results from a search query with multiple indices
+        returning only the latest version of the documents
+        """
+        map_ids = {}
+        for hit in sorted(
+            hits,
+            key=lambda x: datetime.fromisoformat(x["_source"]["timestamp"]),
+            reverse=True,
+        ):
+            if hit["_id"] not in map_ids:
+                map_ids[hit["_id"]] = hit["_source"]["vector_dump"]
+
+        return map_ids
+
+    def mget(self, keys: Sequence[str]) -> List[Optional[List[float]]]:
+        """Get the values associated with the given keys."""
+        if not any(keys):
+            return []
+
+        cache_keys = [self._key(k) for k in keys]
+        if self._is_alias:
+            try:
+                results = self._es_client.search(
+                    index=self._index_name,
+                    body={
+                        "query": {"ids": {"values": cache_keys}},
+                        "size": len(cache_keys) * self._maximum_duplicate_allowed,
+                    },
+                    source_includes=["vector_dump", "timestamp"],
+                )
+
+            except BadRequestError as e:
+                if "window too large" in (
+                    e.body.get("error", {}).get("root_cause", [{}])[0].get("reason", "")
+                ):
+                    logger.warning(
+                        "Exceeded the maximum window size, "
+                        "Reduce the duplicates manually or lower "
+                        "`maximum_duplicate_allowed.`"
+                    )
+                    raise e
+
+            total_hits = results["hits"]["total"]["value"]
+            if self._maximum_duplicate_allowed > 1 and total_hits > len(cache_keys):
+                logger.warning(
+                    f"Deduplicating, found {total_hits} hits for {len(cache_keys)} keys"
+                )
+                map_ids = self._deduplicate_hits(results["hits"]["hits"])
+            else:
+                map_ids = {
+                    r["_id"]: r["_source"]["vector_dump"]
+                    for r in results["hits"]["hits"]
+                }
+
+            return [map_ids.get(k) for k in cache_keys]
+
+        else:
+            records = self._es_client.mget(
+                index=self._index_name, ids=cache_keys, source_includes=["vector_dump"]
+            )
+            return [
+                r["_source"]["vector_dump"] if r["found"] else None
+                for r in records["docs"]
+            ]
+
+    def build_document(self, text_input: str, vector: List[float]) -> Dict[str, Any]:
+        """Build the Elasticsearch document for storing a single embedding"""
+        body: Dict[str, Any] = {
+            "vector_dump": vector,
+            "timestamp": datetime.now().isoformat(),
+        }
+        if self._metadata is not None:
+            body["metadata"] = self._metadata
+        if self._store_input:
+            body["text_input"] = text_input
```

Comment:
I assume you want to store these so you can inspect the cache and learn something about usage statistics? Wondering if this is the right way to do this (or rather collect stats differently), but I'm not opposed to it

Reply:
The core idea behind storing both the text and embeddings together is to ensure that we have the possibility to reconstruct the model or utilize them in other ways in the future. It is just to preserve valuable information for potential future use cases.